File: apps/api/app/schemas/remediation.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from ..models import RemediationActionStatus, RemediationPlanStatus, RemediationVerificationStatus


class RemediationGenerate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    recommendation_id: UUID


class RemediationActionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    sequence: int
    action_key: str
    parameters: dict
    status: RemediationActionStatus
    result: dict | None


class RemediationVerificationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    device_id: UUID
    check_key: str
    status: RemediationVerificationStatus
    observed: dict | None
    details: str | None
    verified_at: datetime | None


class RemediationPlanRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    organization_id: UUID
    device_id: UUID
    recommendation_id: UUID | None
    root_cause_finding_id: UUID | None
    created_by_user_id: UUID | None
    approved_by_user_id: UUID | None
    title: str
    rationale: str
    plan_hash: str
    dry_run: bool
    status: RemediationPlanStatus
    verification_status: RemediationVerificationStatus
    approved_at: datetime | None
    executed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    actions: list[RemediationActionRead] = Field(default_factory=list)
    verifications: list[RemediationVerificationRead] = Field(default_factory=list)


class RemediationPlanList(BaseModel):
    items: list[RemediationPlanRead]
    total: int

File: apps/web/src/types/remediation.ts
/** apps/api/app/schemas/remediation.py::RemediationActionRead */
export type RemediationAction = {
  id: string
  sequence: number
  action_key: string
  parameters: Record<string, unknown>
  status: 'pending' | 'ready' | 'executing' | 'succeeded' | 'failed' | 'skipped' | 'cancelled'
  result: Record<string, unknown> | null
}

/** apps/api/app/schemas/remediation.py::RemediationVerificationRead */
export type RemediationVerification = {
  id: string
  device_id: string
  check_key: string
  status: 'pending' | 'passed' | 'failed' | 'inconclusive'
  observed: Record<string, unknown> | null
  details: string | null
  verified_at: string | null
}

/** apps/api/app/schemas/remediation.py::RemediationPlanRead */
export type RemediationPlan = {
  id: string
  organization_id: string
  device_id: string
  recommendation_id: string | null
  root_cause_finding_id: string | null
  title: string
  rationale: string
  plan_hash: string
  dry_run: boolean
  status:
    | 'draft'
    | 'pending_approval'
    | 'approved'
    | 'rejected'
    | 'queued'
    | 'executing'
    | 'succeeded'
    | 'failed'
    | 'cancelled'
    | 'verification_required'
    | 'verified'
  verification_status: 'pending' | 'passed' | 'failed' | 'inconclusive'
  approved_at: string | null
  executed_at: string | null
  created_at: string
  updated_at: string
  actions: RemediationAction[]
  verifications: RemediationVerification[]
}
